### DISCUSSION TITLE
read_cb behavior after read_stop

### DISCUSSION THREAD

**Discussion by rexroni** [libuv/libuv#3740] [Q&A] [answered]
I am trying to understand the semantics of the `read_cb` when `read_stop` is called.

The docs say:

> Stop reading data from the stream. The uv_read_cb callback will no longer be called.

But I don't see how not calling `read_cb` again can be guaranteed, when `alloc_cb` may be used to [preallocate a buffer](https://github.com/libuv/libuv/blob/v1.x/src/win/tcp.c#L493) on windows that goes into a queue for later processing.  I'll admit that I don't totally understand the windows overlapped IO API so I could just be missing something obvious.

But what happens after a quick sequence of `read_start()`, `read_stop()`, `read_start()`?  Could a buffer allocated by the first alloc_cb end up being passed into the second read_cb?

**Comment by vtjnash** [comment, marked as answer]
That is dead code and never run

**Reply by rexroni** [reply]
Ok.  I see the comment now:
```c
/*
 * Threshold of active tcp streams for which to preallocate tcp read buffers.
 * (Due to node slab allocator performing poorly under this pattern,
 *  the optimization is temporarily disabled (threshold=0).  This will be
 *  revisited once node allocator is improved.)
 */
const unsigned int uv_active_tcp_streams_threshold = 0;
```

At least ostensibly it reads as if I shouldn't count on this behavior forever.  But also this comment is 12 years old.

Can I trust that the current behavior won't change?  And if so, why not just delete the dead code?

From my perspective as a libuv user, I don't want to write code that might suddenly start segfaulting on windows in 2 or 3 years, so if the code says this situation is temporary I am weary to depend on the current behavior.

**Reply by bnoordhuis** [reply]
> Stop reading data from the stream. The uv_read_cb callback will no longer be called.

Looks like a lacuna both in libuv's documentation and its implementation.

When you call `uv_read_stop()` inside or after `uv_alloc_cb` but before `uv_read_cb`, then libuv should call `uv_read_cb` with `nread==0` to indicate it's giving you your buffer back.

Instead it segfaults because `uv_read_stop()` clears the `uv_read_cb` pointer.

**Reply by bnoordhuis** [reply]
Apropos the TCP (and UDP) zero read code in src/win: I would consider that dead after ~11 years and a good candidate for removal. :-)

**Reply by bnoordhuis** [reply]
https://github.com/libuv/libuv/pull/3743

**Reply by vtjnash** [reply]
It is already deleted on master, it just wasn't deemed worth cleaning up in the v1.x branch at the time. It would be an API break to enable it anyways for that reason.
